lib/mpl_toolkits/mplot3d/proj3d.py

```diff
@@ -146,6 +144,31 @@ def proj_transform(xs, ys, zs, M):
 transform = proj_transform
 
 
+def proj_transform_vectors(vecs, M):
+    """Vector version of ``project_transform`` able to handle MaskedArrays.
+
+    Parameters
+    ----------
+    vecs : 3xN np.ndarray or np.ma.MaskedArray
```

Comment:
It might make more sense for this to be a `(..., 3)` array, which would make the below...

Reply:
Note that those reshapes should not trigger any reallocations, so I'm not sure if that would actually help. We could push the transposes into this function, but we cannot make them go away without changing the conventions around `renderer.M`. Still, I don't think they cause any harm here.

Reply:
This suggestion is more based off simplicity and convention than it is performance.

Reply:
I've tried your suggestions and it seems like while the original version takes ~72ms on average, the proposed changes increase the runtime to ~127ms. This is likely due to strided memory patters that start pop up everywhere (padding the array, normalizing the product, reductions over z axis). I do agree that the code is nicer though.

Reply:
Thanks for profiling. Is that the runtime of the entire plot, or just this function?

Reply:
It's the runtime for the part of this function that does array math. Measuring the time for the complete plotting process has too much variability, so it's difficult to actually see the differences. Still, as much as microseconds don't matter to me, 50ms seems like a difference that we should be taking into account (stack a few of those and you get a significant fraction of a second!).

Reply:
Can you put a complete benchmark in a gist so that I can play around?

> This is likely due to strided memory patters that start pop up everywhere (padding the array, normalizing the product, reductions over z axis)

It's possible to eliminate this striding by changing the memory order. `order='F'` in the `np.empty` call might do the job.

What numpy version are you timing with?

Reply:
Uh sure, I'll try to isolate something. Good point about `order='F'`. I'm using MKL-based NumPy from conda (version 1.18.1) on a 2012 MBP.

Reply:
[Here you go](https://gist.github.com/apaszke/fef347c22472c12f0cfb9da17e8241eb). Outputs this on my laptop:
```
New:		 92.46ms
Baseline:	 46.31ms
Diff:		-46.15ms
```

Reply:
Fantastic, thanks for assembling that.

Reply:
Seems that something screwy is going on with `matmul(A, B)` vs `matmul(B.T, A.T).T` for 2d `A` and `B`. `matmul` is supposed to know they are equivalent and pick the faster one, but it seems to be doing the same thing for both.

Reply:
How large part of the slowdown is caused by matmul picking the wrong/having a slow kernel for this case then?

Reply:
I can get a speed _up_ if with this mess:
```
def _proj_transform_vectors_new(vecs, M):
    is_masked = isinstance(vecs, np.ma.MaskedArray)
    if is_masked:
        mask = vecs.mask
    # allocate with a convenient memory order
    vecs_pad = np.moveaxis(np.empty((vecs.shape[-1] + 1,) + vecs.shape[:-1]), 0, -1)
    vecs_pad[..., :-1] = vecs
    vecs_pad[..., -1] = 1
    product = np.empty_like(vecs_pad)
    np.matmul(vecs_pad, M.T, out=product)
    result = product[..., :3] / product[..., 3:]
    if is_masked:
        return np.ma.array(result, mask=mask)
    else:
        return result


def new():
    # reshape is for speed only, it works correctly without
    psegments = _proj_transform_vectors_new(segments.reshape(-1, 3), M).reshape(segments.shape)
    epilogue(psegments)
    return psegments
``

Reply:
This indeed yields a speedup, but a tiny one on my laptop:
```
New:		 42.17ms
Baseline:	 45.15ms
Diff:		 2.98ms
```

I'm ok with going either way, just let me know 🤷‍♂

Reply:
I think my request would perhaps be just to move the reshapes you have at the call site inside the function - that way you can keep the conventional `(...,N)` API, but the implementation can be the easy-to-read and efficient code you have now.